### ISSUE TITLE
keyword oriented matchfeatures are sometimes 0  

### ISSUE THREAD

**Issue by jobergum** [vespa-engine/vespa#31442]
## TL;DR

keyword oriented features calculated in matchfeatures are incorrect 

### Background

Encountered this while working on introducing a new rank-profile for [search.vespa.ai](search.vespa.ai) where the returned match features are set to 0 for keyword matches, while summaryfeatures returns the expected values. It looks to be related to the weakAnd targetHits, as setting a higher targetHits than 100 makes the problem go away. 

### Details

The  hybrid query combining two embedding fields and regular text search, plus a rank term on a exact matched field. 

```
[RANK (OR (WEAKAND(100) batch delete many document) NEAREST_NEIGHBOR {field=embedding,queryTensorName=q,hnsw.exploreAdditionalHits=0,distanceThreshold=Infinity,approximate=true,targetHits=100} NEAREST_NEIGHBOR {field=question_embedding,queryTensorName=q,hnsw.exploreAdditionalHits=0,distanceThreshold=Infinity,approximate=true,targetHits=100}) questions_exact:batch delete many documents] timeout=491ms offset=0 hits=32 rankprofile[hybrid-with-clamp]
```

The `hybrid-with-clamp` rank profile: 

```batch
rank-profile hybrid-with-clamp {
        inputs {
            query(q) tensor<float>(x[384])
            query(semantic_weight) double: 0.6
        }
        function scale(val) {
            expression: 2*atan(val/4)/(3.14159)
        }

        function scaled_bm25_content() {
            expression: scale(bm25(content))
        }

        function scaled_bm25_title() {
            expression: scale(bm25(title))
        }

        function scaled_bm25_path() {
            expression: scale(bm25(path))
        }

        function combined_bm25() {
            expression: 0.3*scale(bm25(title)) + 0.6*scale(bm25(content)) + 0.1*scale(bm25(path))
        }

         function semantic_content() {
            expression: cos(distance(field, embedding))
         }

         function semantic_question() {
            expression: max(cos(distance(field, question_embedding)),0)
         }

         function semantic_combined() {
            expression: 0.5*semantic_content + 0.5*semantic_question
         }

         function keywords() {
            expression: ( 0.8*combined_bm25  + 0.1*elementCompleteness(questions).completeness) + 0.1*elementCompleteness(questions_exact).completeness
         }

         first-phase {
            expression {
                if(keywords > 0.1, query(semantic_weight)*semantic_combined + (1 - query(semantic_weight))*keywords, -1)
            }
            rank-score-drop-limit: 0.0
         }
         match-features {
            queryTermCount
            keywords
            bm25(title)
            scaled_bm25_title
            bm25(content)
            scaled_bm25_content
            bm25(path)
            scaled_bm25_path
            combined_bm25
            semantic_question
            semantic_content
            semantic_combined
            elementCompleteness(questions).completeness
            elementCompleteness(questions_exact).completeness
         }
         summary-features {
            queryTermCount
            keywords
            bm25(title)
            scaled_bm25_title
            bm25(content)
            scaled_bm25_content
            bm25(path)
            scaled_bm25_path
            combined_bm25
            semantic_question
            semantic_content
            semantic_combined
            elementCompleteness(questions).completeness
            elementCompleteness(questions_exact).completeness
         }
     }
```

The above hit is one example of a hit retrieved by the query where `matchfeatures` are 0 for keyword oriented features while `relevance` is correctly computed and `summaryfeatures` are also correct. 0.6*semantic_combined + 0.4*keywords:

```
0.6* 0.829460919862687 + 0.4*0.5484951136389049 = .7170745973731739
```
Which aligns with the computed `relevance` which is assigned by the matching protocol phase. 

```json
 {
                "id": "id:open-p:paragraph::open/en/onnx.html-batch-dimensions",
                "relevance": 0.7170745973731741,
                "source": "documentation",
                "fields": {
                    "matchfeatures": {
                        "bm25(content)": 0.0,
                        "bm25(path)": 0.0,
                        "bm25(title)": 0.0,
                        "elementCompleteness(questions).completeness": 0.0,
                        "elementCompleteness(questions_exact).completeness": 0.0,
                        "queryTermCount": 7.0,
                        "combined_bm25": 0.0,
                        "keywords": 0.0,
                        "scaled_bm25_content": 0.0,
                        "scaled_bm25_path": 0.0,
                        "scaled_bm25_title": 0.0,
                        "semantic_combined": 0.829460919862687,
                        "semantic_content": 0.8072345001900495,
                        "semantic_question": 0.8516873395353245
                    },
                    "sddocname": "paragraph",
                    "doc_id": "/en/onnx.html",
                    "documentid": "id:open-p:paragraph::open/en/onnx.html-batch-dimensions",
                    "path": "/en/onnx.html#batch-dimensions",
                    "base_uri": "https://docs.vespa.ai",
                    "title": "Ranking With ONNX Models -  Batch dimensions",
                    "content": "When training your model you will typically have an input that contains a\ndimension for batches, for instance an input with sizes `[-1, 784]`. Here, -1\ntypically denotes the batch dimension. During ONNX inference in ranking, Vespa uses batch size 1.",
                    "questions": [
                        "What does -1 denote?",
                        "What is batch dimension?",
                        "What is batch size?",
                        "What is mini-batch?",
                        "What is Vespa run-time evaluation?",
                        "How does Vespa do inference?",
                        "What should batch dimension be set to during run-time evaluation?",
                        "Why set batch dimension to 1 during run-time evaluation?"
                    ],
                    "content_tokens": 58,
                    "namespace": "open-p",
                    "selfhosted": false,
                    "summaryfeatures": {
                        "bm25(content)": 8.29638549501862,
                        "bm25(path)": 5.679345860036248,
                        "bm25(title)": 4.685141766103279,
                        "elementCompleteness(questions).completeness": 0.25,
                        "elementCompleteness(questions_exact).completeness": 0.0,
                        "queryTermCount": 7.0,
                        "combined_bm25": 0.654368892048631,
                        "keywords": 0.5484951136389049,
                        "scaled_bm25_content": 0.7139954101063665,
                        "scaled_bm25_path": 0.6093640238364424,
                        "scaled_bm25_title": 0.5501174786705565,
                        "semantic_combined": 0.829460919862687,
                        "semantic_content": 0.8072345001900495,
                        "semantic_question": 0.8516873395353245,
                        "vespa.summaryFeatures.cached": 0.0
                    }
                }
 }
```

**Comment by luoyetx** [comment]
we meet the similar problem with nearest neighbor query, because the [heap](https://github.com/vespa-engine/vespa/blob/master/searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_iterator.cpp#L35) used by the iterator is not reset when fetching summaries, this cause the hit from search phase may not be considered as a hit in summary phase.

**Comment by luoyetx** [comment]
this also happens in second phase ranking, but the heap is used by multi threads, not easy to reset it once during the second phase ranking

> we meet the similar problem with nearest neighbor query, because the [heap](https://github.com/vespa-engine/vespa/blob/master/searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_iterator.cpp#L35) used by the iterator is not reset when fetching summaries, this cause the hit from search phase may not be considered as a hit in summary phase.

**Comment by baldersheim** [comment]
That is correct, resolution has been designed and will appear in a not too distant future.
Note that this has been an issue with nearest neighbor and wand, also known as parallell wand since their inception.
This became a larger problem after we started using the same heap for the general weakand.
